packages/vue-router/src/createApp.ts

```diff
@@ -99,15 +105,19 @@ export const createApp = (
                         path.replace(/\/$/, "") === to.path.replace(/\/$/, "")
                     );
 
-                    const nextPathname = isToAppRoute
-                      ? dedupeSlash([basename, to.path].join("/"))
-                      : to.path;
-
-                    onRemoteNavigate?.({
-                      pathname: nextPathname,
-                      hash: to.hash,
-                      // TODO search: to.query,
-                    });
+                    if (!isToAppRoute) {
+                      onRemoteNavigate?.({
+                        pathname: to.path,
+                        hash: to.hash,
+                        // TODO search: to.query,
+                      });
+                    } else {
+                      const nextPathname = dedupeSlash(
+                        [basename, to.path].join("/")
+                      );
+                      // update url and navigation history
+                      window.history.pushState({}, "", nextPathname);
```

Comment:
We are using a `history` abstraction on top of window history. We should not manipulate the window directly because it can create race conditions or strange behaviors. Also, window will break when doing server-side rendering. I think it's better we remove the performance optimization for now unless we can make it work without manipulating window history.